File: backend/src/main/java/com/back/domain/pet/entity/Pet.java
package com.back.domain.pet.entity;

import com.back.domain.adoption.entity.Adoption;
import com.back.domain.care.entity.Care;
import com.back.domain.member.entity.Member;
import com.back.domain.pet.enums.Gender;
import com.back.domain.pet.enums.PetStatusType;
import com.back.domain.shelter.entity.Shelter;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.Lob;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import java.util.ArrayList;
import java.util.List;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

@Entity
@Getter
@Table(name = "pet")
@EntityListeners(AuditingEntityListener.class)
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Pet {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "pet_id")
    private Long id;

    @Column(name = "pet_name", nullable = false)
    private String name;

    @Column(name = "pet_species", nullable = false)
    private String species;

    @Column(name = "pet_age")
    private int age;

    @Column(name = "pet_gender", nullable = false)
    @Enumerated(EnumType.STRING)
    private Gender gender;

    @Lob
    @Column(name = "pet_description")
    private String description;

    @Column(name = "pet_image_url")
    private String imageUrl;

    @OneToMany(mappedBy = "pet", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PetStatus> petStatuses = new ArrayList<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shelter_id")
    private Shelter shelter;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id")
    private Member member;

    @OneToMany(mappedBy = "pet", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PetStatus> statuses = new ArrayList<>();

    @OneToMany(mappedBy = "pet", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Adoption> adoptions = new ArrayList<>();

    @OneToMany(mappedBy = "pet", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Care> cares = new ArrayList<>();

    @Builder
    public Pet(Long id, String name, String species, int age, Gender gender, List<PetStatus> petStatuses,
               String description, String imageUrl, Shelter shelter, Member member) {
        this.id = id;
        this.name = name;
        this.species = species;
        this.age = age;
        this.gender = gender;
        this.description = description;
        this.imageUrl = imageUrl;
        this.shelter = shelter;
        this.member = member;
        this.petStatuses = petStatuses;
    }

    public void setPetStatuses(List<PetStatus> testPetStatus) {
        this.petStatuses = testPetStatus;
    }
}

File: backend/src/main/java/com/back/domain/pet/entity/Pet.kt
package com.back.domain.pet.entity

import com.back.domain.adoption.entity.Adoption
import com.back.domain.care.entity.Care
import com.back.domain.member.entity.Member
import com.back.domain.pet.dto.request.PetUpdateRequestDto
import com.back.domain.pet.enums.Gender
import com.back.domain.shelter.entity.Shelter
import jakarta.persistence.*
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.LocalDateTime

@Entity
@Table(name = "pet")
@EntityListeners(AuditingEntityListener::class)
class Pet protected constructor(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "pet_id")
    val id: Long? = null,

    @Column(name = "pet_name", nullable = false)
    var name: String,

    @Column(name = "pet_species", nullable = false)
    var species: String,

    @Column(name = "pet_age")
    var age: Int,

    @Enumerated(EnumType.STRING)
    @Column(name = "pet_gender", nullable = false)
    var gender: Gender,

    @Lob
    @Column(name = "pet_description")
    var description: String? = null,

    @Column(name = "pet_image_url")
    var imageUrl: String? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shelter_id", nullable = true)
    var shelter: Shelter? = null,

) {
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", nullable = false)
    lateinit var member: Member

    @CreatedDate
    @Column(updatable = false)
    var createdAt: LocalDateTime? = null
        protected set

    @OneToMany(mappedBy = "pet", cascade = [CascadeType.ALL], orphanRemoval = true)
    val adoptions: MutableList<Adoption> = mutableListOf()

    @OneToMany(mappedBy = "pet", cascade = [CascadeType.ALL], orphanRemoval = true)
    var petStatuses: MutableList<PetStatus> = mutableListOf()

    @OneToMany(mappedBy = "pet", cascade = [CascadeType.ALL], orphanRemoval = true)
    val cares: MutableList<Care> = mutableListOf()


    fun updatePet(dto: PetUpdateRequestDto) {
        dto.name?.let { this.name = it }
        dto.species?.let { this.species = it }
        dto.age?.let { this.age = it }
        dto.gender?.let { this.gender = it }
        dto.description?.let { this.description = it }
        dto.imageUrl?.let { this.imageUrl = it }
    }
    companion object {
        fun create(
            name: String,
            species: String,
            age: Int,
            gender: Gender,
            description: String? = null,
            imageUrl: String? = null,
            shelter: Shelter? = null,
            member: Member,
            statuses: List<PetStatus> = emptyList()
        ): Pet {
            val pet = Pet(
                name = name,
                species = species,
                age = age,
                gender = gender,
                description = description,
                imageUrl = imageUrl,
                shelter = shelter,
            )
            pet.member = member
            pet.petStatuses = statuses.toMutableList()
            return pet
        }
    }
}